### PR TITLE
Add Mutex::arc and RwLock::arc

### LINKED ISSUE #74866: Add Mutex::arc method
Mutexes are seldom used without an Arc, I propose to add a method `Mutex::arc` that returns an `Arc<Mutex<_>>`. This would make code more readable and compact almost everywhere mutexes are used. It also makes the intent clearer, as we first think about the `Mutex`, not the `Arc` when we use one.

```rust
// this:
let mutex = Arc::new(Mutex::new(Thing { inner: 0 }));
// would turn into this
let mutex = Mutex::arc(Thing { inner: 0 });
```

 This would also mimic what we already have with the `Box::pin` method that return a `Pin<Box<_>>`.

### PR DESCRIPTION
**Pull request by Stupremee**
This PR introduces the `arc` function for `RwLock` and `Mutex`.
`Mutex::arc`/`RwLock::arc` is a shorthand for `Arc::new(Mutex::new(t))`

Resolves #74866

### CODE REVIEW

File: library/std/src/lib.rs [hunk 1]
```diff
@@ -282,6 +282,7 @@
 #![feature(maybe_uninit_ref)]
 #![feature(maybe_uninit_slice)]
 #![feature(min_specialization)]
+#![feature(mutex_arc)]
```
Comment (LeSeulArtichaut) [inline library/std/src/lib.rs:285]: Is this only used for tests? If so, maybe:
```suggestion
#![cfg_attr(feature(mutex_arc))]
```
?

**Comment by sdroege** [comment]
Should there also be a `RefCell::rc()` then? That's approximately the same as `RWLock::arc()` in a single-threaded context

**Comment by Stupremee** [comment]
> Should there also be a `RefCell::rc()` then? That's approximately the same as `RWLock::arc()` in a single-threaded context

Yes that's a good idea

**Comment by dtolnay** [comment]
I think I would prefer not to add these but I will ping the library team to give a chance to jump in and make a case if someone else is interested.

@rust-lang/libs:
@rfcbot fcp close

This PR is proposing `Mutex::arc` and `RwLock::arc` constructors that are equivalent to:

```diff
- let arc = Arc::new(Mutex::new(value));
+ let arc = Mutex::arc(value);
```

There is some consistency here with [`Box::pin`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.pin) which returns Pin\<Box\<T\>\>, but here I'm not convinced of the need for these over Arc::new(Mutex::new(...)).

**Comment by sfackler** [comment]
Box::pin is a different case because you can't otherwise construct the type without `unsafe { ... }`. In this case the proposed functions are just there to save a couple characters which does not seem worth it IMO.

**Review by dtolnay — COMMENTED** [review]
I'll go ahead and close this. Thanks anyway @Stupremee!

**Pull request closed without merging**